indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java

```diff
@@ -78,164 +85,354 @@ public WorkerTaskMonitor(
               );
 
-              if (taskStatusAlreadySubmitted != null) {
-                log.info("Affirmative. Connecting to already-running task [%s]", task.getId());
-                taskStatus = taskStatusAlreadySubmitted.get();
-              } else {
-                log.info("Affirmative. Running task [%s]", task.getId());
-                workerCuratorCoordinator.unannounceTask(task.getId());
-                taskStatus = taskRunner.run(task).get();
-              }
-            }
-            catch (InterruptedException e) {
-              log.debug(e, "Interrupted while running task[%s], exiting.", task.getId());
-              return;
-            }
-            catch (Exception e) {
-              log.makeAlert(e, "I can't build there. Failed to run task")
-                 .addData("task", task.getId())
-                 .emit();
-              taskStatus = TaskStatus.failure(task.getId());
-            }
-            finally {
-              running.remove(task);
+              notices.add(new RunNotice(task));
             }
+          }
+        }
+    );
+  }
+
+  private void registerLocationListener()
+  {
+    taskRunner.registerListener(
+        new TaskRunnerListener()
+        {
+          @Override
+          public void locationChanged(final String taskId, final TaskLocation newLocation)
+          {
+            notices.add(new LocationNotice(taskId, newLocation));
+          }
+        },
+        MoreExecutors.sameThreadExecutor()
+    );
+  }
 
-            taskStatus = taskStatus.withDuration(System.currentTimeMillis() - startTime);
+  private void addRunningTask(final Task task, final ListenableFuture<TaskStatus> future)
+  {
+    running.put(task.getId(), new TaskDetails(task, future));
+    Futures.addCallback(
+        future,
+        new FutureCallback<TaskStatus>()
+        {
+          @Override
+          public void onSuccess(TaskStatus result)
+          {
+            notices.add(new StatusNotice(task, result));
+          }
 
-            try {
-              workerCuratorCoordinator.updateAnnouncement(TaskAnnouncement.create(task, taskStatus));
-              log.info(
-                  "Job's finished. Completed [%s] with status [%s]",
-                  task.getId(),
-                  taskStatus.getStatusCode()
-              );
-            }
-            catch (Exception e) {
-              log.makeAlert(e, "Failed to update task status")
-                 .addData("task", task.getId())
-                 .emit();
-            }
+          @Override
+          public void onFailure(Throwable t)
+          {
+            notices.add(new StatusNotice(task, TaskStatus.failure(task.getId())));
           }
         }
     );
   }
 
-  private boolean isTaskRunning(final String taskId)
+  @LifecycleStop
+  public void stop()
   {
-    for (final Task runningTask : running) {
-      if (runningTask.getId().equals(taskId)) {
-        return true;
+    synchronized (this) {
```

Comment:
I didn't think it mattered that much as this is not an object that anyone is going to be holding and trying to synchronize on. but yeah, it can be internal.